lmdeploy/pytorch/engine/request.py

```diff
@@ -145,21 +150,13 @@ def run_until_complete(self, future: Awaitable):
 
     def _resp_get(self):
         """resp_que.get."""
-        timeout = 1
-
-        while True:
-            try:
-                ret = self.resp_thread_que.get(timeout=timeout)
-                return ret
-            except Empty:
-                if not self.manager.is_loop_alive():
-                    logger.debug('Engine loop is not alive.')
-                    exit(1)
-                continue
-            except Exception as e:
-                logger.exception(
-                    f'sender[{self.sender_id}] get response failed: {e}')
-                raise e
+        timeout = 1.0
+        que = self.resp_thread_que
+        not_empty = que.not_empty
+        with not_empty:
+            while not que._qsize():
+                not_empty.wait(timeout)
+        return que.get_nowait()
```

Comment:
Does this `while not que._qsize():` also block the thread?

Reply:
No, `_qsize()` is `len(que.queue)`
https://github.com/python/cpython/blob/733dac01b0dc3047efc9027dba177d7116e47c50/Lib/queue.py#L267

and `que.queue` is just a `deque.queue`
https://github.com/python/cpython/blob/733dac01b0dc3047efc9027dba177d7116e47c50/Lib/queue.py#L264-L265